### ISSUE TITLE
Problem in restarting chains

### ISSUE THREAD

**Issue by frompineve** [brinckmann/montepython_public#116]
Hi everyone,

I am facing the following problem with the most recent version of montepython. After running montepython for couple of days, I need to stop the runs and restart them.
I use the suggested command:

srun python montepython/MontePython.py run --restart chains/test/2020-03-11_10000000__1.txt -N 10000001 

OR also

srun python montepython/MontePython.py run -r chains/test/2020-03-11_10000000__1.txt -N 10000001 

Both ways, I always get the following error after the montepython crashes almost immediately:

Traceback (most recent call last):
File "montepython/MontePython.py", line 40, in <module>
sys.exit(run())
File "/cluster/home/frompi01/montepython/montepython/run.py", line 45, in run
sampler.run(cosmo, data, command_line)
File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 46, in run
mcmc.chain(cosmo, data, command_line)
File "/cluster/home/frompi01/montepython/montepython/mcmc.py", line 455, in chain
sampler.read_args_from_chain(data, command_line.restart)
File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 106, in read_args_from_chain
if chain_file.tail(1)[0].split('\t')[0] == '#':
IndexError: list index out of range

Has anybody maybe faced this already? How do you normally restart chains? 
Thank you for the help!
Best

Fabrizio

**Comment by brinckmann** [comment]
Hi Fabrizio,

The syntax for restarting is correct. As for the problem, I have occasionally noticed empty lines in my chains files. I haven't figured out where these come from, but they interfere with analyzing chains and I would not be surprised if it also interferes with restarting, so it may be the same problem.

Can you check if you can analyze the chains with info? If not, you should be able to guess which of your chains files has a problem. Then maybe you can find the problem in the file and manually fix it by deleting the empty line, but if not you should be able to simply exclude the file from the analysis/restart with some manual renaming of files.

A better solution would be to try to add a bugfix that simply ignores empty lines when restarting or analyzing (or better yet, figure out how to stop them from occasionally occurring), but I don't have time to figure this out now.

I hope this helps!

Best,
Thejs

**Comment by frompineve** [comment]
Hi Thejs,

thanks a lot for the help. The info works smoothly. It is only when I try to restart the chains that things do not work. Unfortunately this occurs for all the runs that I do. 
Thanks,
Best
Fabrizio

**Comment by brinckmann** [comment]
Which version of python are you using?

**Comment by frompineve** [comment]
3.6.0

**Comment by brinckmann** [comment]
I'm not very familiar with Python 3, can you guess if the line in question has a bug from the conversion to Python 3 compatibility?
I.e. line 105-106 montepython/montepython/sampler.py:
```
    while not success:
        if chain_file.tail(1)[0].split('\t')[0] == '#':
```
If one of the chain files end with lines that are commented with a "#", it is supposed to find a line in the chain file that is not commented. It works for Python 2.7, but obviously we want it working for Python 3 as well. Perhaps @schoeneberg or @fkoehlin want to chip in?

As it is supposed to catch unlikely events where the last line is commented you could probably just comment lines 105-112.

**Comment by schoeneberg** [comment]
Dear frompineve,

could you post also the last few lines of the chain file that you are restarting? It would be very helpful! I think what might be an issue is that there is no '\t' (tab character) within your last line.

@brinckmann , shouldn't the code be searching for 

if chain_file.tail(1)[0][0] == '#':

instead? Why do you require there to be a tab character as well?

**Comment by frompineve** [comment]
Hi schoeneberg,

thank you and brinckmann for helping with this! I am not particularly familiar with python, therefore I would not know why the restart is not working on my chains. Today I tried again on other chains, which I stopped after 1 day (I am working on a local cluster with python 3.6.0 and the latest version of montepython) and got the following error message

Traceback (most recent call last):
  File "montepython/MontePython.py", line 40, in <module>
    sys.exit(run())
  File "/cluster/home/frompi01/montepython/montepython/run.py", line 45, in run
    sampler.run(cosmo, data, command_line)
  File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 46, in$
    mcmc.chain(cosmo, data, command_line)
  File "/cluster/home/frompi01/montepython/montepython/mcmc.py", line 455, in c$
    sampler.read_args_from_chain(data, command_line.restart)
  File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 106, i$
    if chain_file.tail(1)[0].split('\t')[0] == '#':
  File "/cluster/home/frompi01/montepython/montepython/io_mp.py", line 483, in $
    lines_found += self.read(1024).count('\n')
TypeError: a bytes-like object is required, not 'str'

I restarted the chains via:

srun python montepython/MontePython.py run --restart chains/test/2020-03-12_10000000__1.txt -N 10000001

The last few lines of the chain above are:

15  505.146	2.243791e+00	1.185739e-01	1.041438e+00	3.057743e+00	9.693253e-01	6.365264e-02	1.000041e+00	8.580429e+00	6.918817e-01	2.478686e-01	6.781319e+01	2.127948e+00	8.115969e-01	
17  505.284	2.243791e+00	1.185739e-01	1.041438e+00	3.057743e+00	9.693253e-01	6.365264e-02	9.994071e-01	8.580429e+00	6.918817e-01	2.478686e-01	6.781319e+01	2.127948e+00	8.115969e-01	
10  505.295	2.243791e+00	1.185739e-01	1.041438e+00	3.057743e+00	9.693253e-01	6.365264e-02	1.000293e+00	8.580429e+00	6.918817e-01	2.478686e-01	6.781319e+01	2.127948e+00	8.115969e-01	

Thank you for the help,
Best

**Comment by schoeneberg** [comment]
Dear @frompineve ,

I have figured out a hotfix that allows re-starting chains again. We might have to go carefully through montepython to see if similar issues could arise anywhere. Anyway, the line to change is (@brinckmann ):

In io_mp.py
line 476 to
        line_list = [line.decode("utf-8") for line in self.readlines()]

This would mean that the default output of tail() is a list of strings, as might be expected. 

If this is not desired, an alternative solution is the following:

In smapler.py
line 106 to
        if chain_file.tail(1)[0].split(b"\t")[0] == b"#":

and line 118 to
            chain_file.tail(commented_line+1)[commented_line].split(b"\t")[i])

and finally in io_mp.py
line 474 to
            lines_found += self.read(1024).count(b"\n")

I think we would prefer the former solution definitely. Also, it appears to work with python2 given the limited testing I've done on that so far.

Cheers!

**Comment by frompineve** [comment]
Hi @schoeneberg,

thanks a lot for your help and sorry for the delay of this reply. 

I tried both of the solutions you proposed. The first did not work (similar errors). The second solution worked (thus I modified both sampler.py and io_mp.py). I just restarted my chains using the same syntax as before.

Thank you so much,
Best

**Comment by schoeneberg** [comment]
Dear @frompineve , dear @brinckmann ,

I realize that the first version of the bugfix was incomplete: Two changes are required in that version:

In io_mp.py
line 474 to
lines_found += self.read(1024).count(b"\n")
and line 476 to
line_list = [line.decode("utf-8") for line in self.readlines()]

Both are required to work in python3 and python2 :)

Cheers!

**Comment by frompineve** [comment]
Hi, 
as I keep using the latest version of montepython, let me point out two further issues which I have found:

1) When using the option --extra, I get the following error:

Traceback (most recent call last):
  File "montepython/MontePython.py", line 40, in <module>
    sys.exit(run())
  File "/cluster/home/frompi01/montepython/montepython/run.py", line 32, in run
    custom_command)
  File "/cluster/home/frompi01/montepython/montepython/run.py", line 191, in safe_initialisation
    cosmo, data, command_line, success = initialise(custom_command)
  File "/cluster/home/frompi01/montepython/montepython/initialise.py", line 59, in initialise
    analyze(command_line)
  File "/cluster/home/frompi01/montepython/montepython/analyze.py", line 92, in analyze
    info = Information(command_line)
  File "/cluster/home/frompi01/montepython/montepython/analyze.py", line 2136, in __init__
    exec(open(command_line.optional_plot_file, plot_file_vars).read())
TypeError: open() argument 2 must be str, not dict

2) When trying to run with a bestfit (i.e. option -b), I get the following error:

Traceback (most recent call last):
  File "montepython/MontePython.py", line 40, in <module>
    sys.exit(run())
  File "/cluster/home/frompi01/montepython/montepython/run.py", line 45, in run
    sampler.run(cosmo, data, command_line)
  File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 46, in run
    mcmc.chain(cosmo, data, command_line)
  File "/cluster/home/frompi01/montepython/montepython/mcmc.py", line 460, in chain
    sampler.read_args_from_bestfit(data, command_line.bf)
  File "/cluster/home/frompi01/montepython/montepython/sampler.py", line 155, in read_args_from_bestfit
    sys.stdout.write('from best-fit file : ', elem, ' = ')
TypeError: write() takes exactly one argument (3 given)

They both seem to be related to python 3 vs python 2, although I am not sure. Would you have any suggestions on how to fix them?
Thanks, best

**Comment by brinckmann** [comment]
Please look through issues before reporting new ones (although I recognize sometimes they can be hard to find if titles weren't cleverly chosen), thanks!

The 1st problem was a mixup of where the `.read()` was placed and was addressed in issue #97 with a solution by Ivan Debono.

The 2nd problem concerns how concatenating strings changed with Python version. There are a few different ways to solve it, but one is described by Florian in issue #117 , personally I'm going to use the %s notation in the official fix, but it shouldn't matter much as long as your fix works with your own python version.

We're putting out a new version, probably this week, which will feature all recent bugfixes (and there are many related to Python 2 to Python 3 compatibility, sorry!) as well as the publicly release H0LiCOW likelihood.

Best,
Thejs

**Comment by frompineve** [comment]
Ok, thanks a lot and sorry for repeating reported issues.
Thanks,
best

**Comment by dchooper** [comment]
Hi,

All python3 issues reported here have been addressed in version 3.4, which has now been released.

Best,
Deanna
